### PR TITLE
Yuqiang/esports create tournament

### PR DESCRIPTION
**Pull request by smallst**
from esports page, adding buttons to access mutiplayer arena 
image
image
for an esports-product owner,
in global team, he can see all tournaments he created, and edit them. top 3 are listed. others collapsed. also he can create tournament with all avaliable arenas.
image
image
in each team, he can see tournaments belongs to this team.
image
when creating tournament, team (clan) is required.
image
image

when a teacher do not have esports product yet. show contact us
image

### CODE REVIEW

**Review by mrfinch — COMMENTED** [review]
`Ends in` text doesn't change on Edit tournament submit.

How does a student play in this tournament? Do we need a name for the tournament?

Seems like once a tournament is created, a user can't delete it. They can only stop it by changing endTime but that doesn't give them an option to create more tournaments since they consume their quota. Is that acceptable?

Will we auto-calculate rankings once tournament is over since we ask for 2 days time during creation or will someone need to do it manually?

File: app/core/store/modules/clans.js [hunk 1]
```diff
@@ -91,6 +114,24 @@ export default {
     async fetchChildClanDetails ({ commit }, { id }) {
       const childClans = await getChildClanDetails(id)
       commit('setClanDetails', { clanId: id, childClans })
+    },
+
+    async fetchTournaments ({ commit }, { clanId }) {
```
[the remark this answers is not in the document]
Comment (smallst) [reply app/core/store/modules/clans.js:119]: yeah that's good. i just think it is in clan module so i skip the forClan. can add back

File: app/schemas/models/user.coffee [hunk 2]
```diff
@@ -428,8 +428,19 @@ _.extend UserSchema.properties,
 
       prepaid: c.objectId(links: [ {rel: 'db', href: '/db/prepaid/{($)}'} ])  # required for type: “course” for legacy compatibility, optional for other types
       productOptions:
-        oneOf: [
-          includedCourseIDs: {type: ['array', 'null']}
+        anyOf: [
+          c.object({}, {propetincludedCourseIDs: {type: ['array', 'null']} }), # course
```
[the remark this answers is not in the document]
Comment (smallst) [reply app/schemas/models/user.coffee:432]: oh looks some mess thing. thank you for find it out!

File: app/views/ladder/components/editTournamentModal.vue [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+import _ from 'lodash'
+import moment from 'moment'
+import { mapGetters } from 'vuex'
+
+import { postTournament, putTournament } from '../../../core/api/tournaments'
+
+import Modal from '../../../components/common/Modal'
+import ClanSelector from '../../landing-pages/league/components/ClanSelector.vue'
+
+const HTML5_FMT_DATETIME_LOCAL = 'YYYY-MM-DDTHH:mm' // moment 1.20+ do have this string but we use 1.19 :joy:
```
[the remark this answers is not in the document]
Comment (smallst) [reply app/views/ladder/components/editTournamentModal.vue:96]: oh actually this is copy from a old file. so i don't really know if we already upgrade the momentjs yet 😂

File: app/views/ladder/components/ladderPanel.vue [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+    },
+    canCreate: {
+      type: Boolean
+    },
+    canEdit: {
+      type: Boolean
+    },
+    clanId: {
+      type: String,
+      default: () => ''
```
[the remark this answers is not in the document]
Comment (smallst) [reply app/views/ladder/components/ladderPanel.vue:91]: jslint report should use function i think

File: app/views/ladder/components/ladderPanel.vue [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+      }
+      return ''
+    },
+    clan () {
+      if (this.tournament) {
+        return this.clanByIdOrSlug(this.tournament.clan)
+      }
+      return {}
+    },
+    time () {
```
[the remark this answers is not in the document]
Comment (smallst) [reply app/views/ladder/components/ladderPanel.vue:114]: 😂  okay

File: app/views/ladder/components/editTournamentModal.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (smallst) [reply app/views/ladder/components/editTournamentModal.vue:96]: okay we already use moment 2...

File: app/views/ladder/MainLadderViewV2.vue [hunk 6]
```diff
@@ -105,28 +202,87 @@ export default {
...
+      } else {
+        newSelectedClan = e.target.value
+      }
+
+      const leagueURL = newSelectedClan ? `/league/ladders/${newSelectedClan}` : '/league/ladders'
+
+      application.router.navigate(leagueURL, { trigger: true })
+    },
+    getCanCreateTournamentNums () {
+      const products = me.activeProducts('esports')
```
[the remark this answers is not in the document]
Comment (smallst) [reply app/views/ladder/MainLadderViewV2.vue:267]: @mrfinch  checked. the other time is happen in server side 😂 so i don't think that's a big issue

**Review by mrfinch — APPROVED** [review]

**Pull request merged**